### PR TITLE
tests: rework ieee802154 drivers

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description

Split from #17789, but with a difference: instead of using symlinks the common functionality to interact with ieee802154 netdev drivers is implemented in its own test_utils module.

### Testing procedure

- Run the tests, you should be able to send an receive Ethernet frames
- Green CI

### Issues/PRs references

Split from #17789
Similar to #17813

### CODE REVIEW

File: sys/include/test_utils/netdev_ieee802154_minimal.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    test_utils_netdev_ieee802154_minimal Minimal netdev Ethernet device processing
+ * @ingroup     sys
+ *
+ * @{
+ * @file
+ * @brief       Provides basic functionalities to interact with an Ethernet networking
```
Comment (leandrolanzieri) [inline sys/include/test_utils/netdev_ieee802154_minimal.h:15]: You mean "IEEE 802.15.4" here

File: sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+void _recv(netdev_t *dev)
+{
+    uint8_t src[IEEE802154_LONG_ADDRESS_LEN], dst[IEEE802154_LONG_ADDRESS_LEN];
+    size_t mhr_len, data_len, src_len, dst_len;
+    netdev_ieee802154_rx_info_t rx_info;
+    le_uint16_t src_pan, dst_pan;
+
+    putchar('\n');
+    /* receive the frame */
+    data_len = dev->driver->recv(dev, _buffer, sizeof(_buffer), &rx_info);
```
Comment (leandrolanzieri) [inline sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c:47]: As pointed in #17813, we should have `ssize_t data_len` and check for errors while receiving (`data_len < 0`)

File: tests/driver_at86rf2xx/main.c [hunk 3]
```diff
@@ -11,119 +11,61 @@
...
-
-    if (!dev_success) {
-        puts("No device could be initialized");
-        return 1;
-    }
-
-    _recv_pid = thread_create(stack, sizeof(stack), THREAD_PRIORITY_MAIN - 1,
-                              THREAD_CREATE_STACKTEST, _recv_thread, NULL,
-                              "recv_thread");
+    puts("Test application for AT86RF2XX ethernet device driver");
```
Comment (leandrolanzieri) [inline tests/driver_at86rf2xx/main.c:56]: IEEE 802.15.4 :-)

File: tests/driver_at86rf215/main.c [hunk 4]
```diff
@@ -11,208 +11,93 @@
...
+        puts("Error initializing devices");
+        return 1;
+    }
 
     /* start the shell */
+    puts("Initialization successful - starting the shell now");
+
     char line_buf[SHELL_DEFAULT_BUFSIZE];
-    shell_run(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
+    shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE);
```
Comment (leandrolanzieri) [inline tests/driver_at86rf215/main.c:100]: This application should keep the extra commands that allow to use device-specific features (such as the batmon)

File: tests/driver_kw2xrf/main.c [hunk 5]
```diff
@@ -1,148 +1,71 @@
...
+        puts("Error initializing devices");
+        return 1;
+    }
 
     /* start the shell */
     puts("Initialization successful - starting the shell now");
 
     char line_buf[SHELL_DEFAULT_BUFSIZE];
-    shell_run(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
+    shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE);
```
Comment (leandrolanzieri) [inline tests/driver_kw2xrf/main.c:68]: Here as well, the specific commands should be kept

**Review by leandrolanzieri — COMMENTED** [review]
Test applications that have device-specific shell commands should keep those. Also, please take a look at the initial `puts` of the test applications, as many say Ethernet driver.

**Comment by MrKevinWeiss** [comment]
OK, I think I got everything.  Rebasing is not so much fun.

I expect that some tests may fail due to size after adding back the extra shell commands.

**Comment by MrKevinWeiss** [comment]
Don't know what is going on with the static tests... It looks like murdock is happy.

Should I squash, is there anything I am missing?

**Comment by MrKevinWeiss** [comment]
Tomorrow I will try to post some tests results (when I have some boards).

**Comment by leandrolanzieri** [comment]
> Note that the `kw2xrf` module requires the NETOPT to get the long and short addresses as they are not saved in the descriptor.

@jia200x do you know if this is a correct / acceptable behaviour? [Documentation](https://doc.riot-os.org/structnetdev__ieee802154__t.html) says:
> Extended structure to hold IEEE 802.15.4 driver state.
> Supposed to be extended by driver implementations. The extended structure should contain all variable driver state.

**Comment by leandrolanzieri** [comment]
BTW this needs rebase

**Comment by jia200x** [comment]
>  It also seems like I have a beacon going every minute that I see.

An IEEE 802.15.4 beacon? :open_mouth: 

> @jia200x do you know if this is a correct / acceptable behaviour? [Documentation](https://doc.riot-os.org/structnetdev__ieee802154__t.html) says:

It's definitely not the correct behavior, although I don't think it breaks anything at this point. All network stacks get this information through the driver anyway (which some map to `netdev_ieee802154_t` directly). But still, nothing guarantees that some code reads the address from the `netdev_ieee802154_t` descriptor.

**Comment by MrKevinWeiss** [comment]
rebased and squashed

**Comment by MrKevinWeiss** [comment]
ping before I have to rebase

**Comment by leandrolanzieri** [comment]
@jia200x would you mind taking a final look?

File: sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c (hunk 2 above)
Comment (jia200x) [reply sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c:47]: +1. If the `recv` function fails, that will trigger a buffer overflow.

**Comment by MrKevinWeiss** [comment]
ping @jia200x comments address, maybe a quick test first?

**Comment by jia200x** [comment]
please squash!

**Comment by MrKevinWeiss** [comment]
ping @leandrolanzieri or @jia200x  quick before more boards get added!!! 😈

**Comment by MrKevinWeiss** [comment]
and we are too late...

**Comment by MrKevinWeiss** [comment]
We were a bit too slow on this one so I had to adapt the `tests/driver_at86rf2xx` to include the changes from #18290 
I tested building with both `CFLAGS='-DAT86RF2XX_SMART_IDLE_LISTENING=0' and `CFLAGS='-DAT86RF2XX_SMART_IDLE_LISTENING=1' and it compiles...  Maybe @maribu can take a quick look at 2f0745010f274ac374e64f86598300cfa5b4199e to make sure it was done right (though I think it strange to use internal functions.)

**Comment by maribu** [comment]
Maybe it makes sense to instead move that test function to the generic test application and just test via the `NETOPT_RANDOM`? E.g. the AT86RF215 also implements that. And for the `at86rf2xx` there is not really a loss in test coverage when only the `NETOPT_RANDOM` interface is tested, as this uses the device-specific random function to get the random.

It would make sense to be a bit more strict and also test the return value of the getopt, not just check that the result actually looks random.

**Comment by MrKevinWeiss** [comment]
> Maybe it makes sense to instead move that test function to the generic test application and just test via the `NETOPT_RANDOM`?

Can we make it another PR after this one? It has been some fight with rebasing many times and this is a pretty big one.


> E.g. the AT86RF215 also implements that.

The current implementation guards it with `AT86RF2XX_SMART_IDLE_LISTENING`, it seems that there are some differences, I don't understand it, I just needed to fix the conflicts. 
```c
#if defined(MODULE_AT86RF233) || defined(MODULE_AT86RFR2)
#ifndef AT86RF2XX_SMART_IDLE_LISTENING
#define AT86RF2XX_SMART_IDLE_LISTENING     (1)
#endif
```

File: sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            break;
+
+        case NETDEV_EVENT_TX_NOACK:
+            puts("No ACK");
+            break;
+
+        default:
+            printf("Event: %d\n", event);
+            break;
+    }
```
Comment (maribu) [inline sys/test_utils/netdev_ieee802154_minimal/netdev_ieee802154_minimal.c:195]:
```suggestion
    case NETDEV_EVENT_ISR:
        event_post(EVENT_PRIO_HIGHEST, &device->event);
        break;

    case NETDEV_EVENT_RX_COMPLETE:
        _recv(dev);
        break;

    case NETDEV_EVENT_TX_COMPLETE:
        puts("Tx complete");
        break;

    case NETDEV_EVENT_TX_COMPLETE_DATA_PENDING:
        puts("Tx complete (with pending data)");
        break;

    case NETDEV_EVENT_TX_MEDIUM_BUSY:
        puts("Medium busy");
        break;

    case NETDEV_EVENT_TX_NOACK:
        puts("No ACK");
        break;

    default:
        printf("Event: %d\n", event);
        break;
    }
```

File: sys/include/test_utils/netdev_ieee802154_minimal.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ * @file
+ * @brief       Provides basic functionalities to interact with an
+ *              IEEE 802.15.4 networking device which implements the
+ *              @ref drivers_netdev_api.
+ *
+ * To use the functionalities, include the module
+ * `USEMODULE += test_utils_netdev_ieee802154_minimal`.
+ * The test application should provide:
+ *  - device initialization, via the implementation of @ref netdev_ieee802154_minimal_init_devs
+ *  - number of devices to test, via the definition of @ref NETDEV_IEEE802154_MINIMAL_NUMOF in `init_dev.h`
```
Comment (maribu) [inline sys/include/test_utils/netdev_ieee802154_minimal.h:23]:
```suggestion
 *  - number of devices to test, via the definition of @ref NETDEV_IEEE802154_MINIMAL_NUMOF
 *    in `init_dev.h`
```

The CI frowns very hard on this > 100 char line (:

File: sys/test_utils/netdev_ieee802154_minimal/shell_commands.c [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    netopt_enable_t enable_val;
+    uint16_t u16_val;
+
+    printf("Iface %3d  HWaddr: ", idx);
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_ADDRESS, tmp, IEEE802154_SHORT_ADDRESS_LEN);
+    l2util_addr_to_str(tmp, IEEE802154_SHORT_ADDRESS_LEN, _addr_str);
+    printf("%s", _addr_str);
+
+    printf(", Long HWaddr: ");
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_ADDRESS_LONG, tmp, IEEE802154_LONG_ADDRESS_LEN);
```
Comment (maribu) [inline sys/test_utils/netdev_ieee802154_minimal/shell_commands.c:60]:
```suggestion
    res = netdev_ieee802154_minimal_get(netdev, NETOPT_ADDRESS_LONG, tmp,
                                        IEEE802154_LONG_ADDRESS_LEN);
```

Again a naughty > 100 char line :)

File: sys/test_utils/netdev_ieee802154_minimal/shell_commands.c [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_IS_WIRED, &u16_val, sizeof(u16_val));
+    if (res < 0) {
+        puts(", wireless");
+    }
+    else {
+        puts(", wired");
+    }
+
+    printf("         ");
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_PRELOADING, &enable_val, sizeof(netopt_enable_t));
```
Comment (maribu) [inline sys/test_utils/netdev_ieee802154_minimal/shell_commands.c:135]:
```suggestion
    res = netdev_ieee802154_minimal_get(netdev, NETOPT_PRELOADING, &enable_val,
                                        sizeof(netopt_enable_t));
```

Same

File: sys/test_utils/netdev_ieee802154_minimal/shell_commands.c [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+    if ((res > 0) && (enable_val == NETOPT_ENABLE)) {
+        printf("  PRELOAD");
+    }
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_AUTOACK, &enable_val, sizeof(netopt_enable_t));
+    if ((res > 0) && (enable_val == NETOPT_ENABLE)) {
+        printf("  AUTOACK");
+    }
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_RAWMODE, &enable_val, sizeof(netopt_enable_t));
```
Comment (maribu) [inline sys/test_utils/netdev_ieee802154_minimal/shell_commands.c:143]:
```suggestion
    res = netdev_ieee802154_minimal_get(netdev, NETOPT_RAWMODE, &enable_val,
                                        sizeof(netopt_enable_t));
```

File: sys/test_utils/netdev_ieee802154_minimal/shell_commands.c [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    if ((res > 0) && (enable_val == NETOPT_ENABLE)) {
+        printf("  RAW");
+    }
+    res = netdev_ieee802154_minimal_get(netdev, NETOPT_AUTOCCA, &enable_val, sizeof(netopt_enable_t));
```
Comment (maribu) [inline sys/test_utils/netdev_ieee802154_minimal/shell_commands.c:147]:
```suggestion
    res = netdev_ieee802154_minimal_get(netdev, NETOPT_AUTOCCA, &enable_val,
                                        sizeof(netopt_enable_t));
```

**Comment by MrKevinWeiss** [comment]
I applied the fix from #18294 in the rebase

File: tests/driver_at86rf2xx/README.md [hunk 12]
```diff
@@ -1,16 +1,16 @@
...
-This is a manual test application for the AT86RF2xx radio driver
+This is a manual test application for the AT86RF2xx radio driver.
 
 For running this test, you need to connect/configure the following pins of your
 radio device:
 - SPI MISO
 - SPI MOSI
 - SPI CLK
-- CS (ship select)
+- CS (chip select)
```
Comment (maribu) [inline tests/driver_at86rf2xx/README.md:9]: ship

**Review by maribu — COMMENTED** [review]
Looks good to me. Thanks for the clean up. Please squash right in the style nit-picks.

File: tests/driver_at86rf215/init_dev.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "at86rf215_params.h"
+#include "kernel_defines.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define AT86RF215_NUM                   ARRAY_SIZE(at86rf215_params)
+#define NETDEV_IEEE802154_MINIMAL_NUMOF AT86RF215_NUM
```
Comment (maribu) [inline tests/driver_at86rf215/init_dev.h:29]:
```suggestion
/* AT86RF215 contains one sub-GHz and one 2.4 GHz driver */
#define NETDEV_IEEE802154_MINIMAL_NUMOF (2 * AT86RF215_NUM)
```

File: tests/driver_at86rf215/main.c [hunk 14]
```diff
@@ -11,38 +11,31 @@
...
-#include "net/gnrc.h"
+#include "thread.h"
+#include "event.h"
+#include "event/thread.h"
+#include "sys/bus.h"
 #include "od.h"
 
-static char batmon_stack[THREAD_STACKSIZE_MAIN];
-static at86rf215_t *dev;
+static at86rf215_t at86rf215[AT86RF215_NUM];
```
Comment (maribu) [inline tests/driver_at86rf215/main.c:36]:
```suggestion
static at86rf215_t at86rf215[NETDEV_IEEE802154_MINIMAL_NUMOF];
```

**Review by maribu — APPROVED** [review]
ACK. Code looks good. I tested with `tests/driver_at86rf2xx` using the `samr21-xpro` and with `tests/driver_at86rf215` using our custom `miot-nucleo-f767zi` testbed board. There is one issue with the AT86RF215 test application, as that needs to allocate twice the number of netdevs, as the sub-GHz and the 2.4 GHz transceiver in that chip are represented as two netdevs in RIOT. I already tested it successfully with the proposed changes.

**Comment by MrKevinWeiss** [comment]
:cry:

**Comment by MrKevinWeiss** [comment]
ping @maribu applied the fixes and rebased.

**Review by maribu — APPROVED** [review]

**Comment by MrKevinWeiss** [comment]
Many thanks @leandrolanzieri @maribu and @jia200x

**Pull request merged**